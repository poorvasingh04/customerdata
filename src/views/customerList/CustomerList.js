import React, { useEffect } from 'react';
import {
  FlatList,
  View,
  Text,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import AppConstants from '../../constants/AppConstants';
import AppStrings from '../../constants/AppStrings';
import { fetchCustomerData } from '../../services/actions';
import distanceFromDublinOffice from '../../utils/DistanceCalculator';
import Customer from './Customer';
import style from './style';

const {
  rowStyle,
  headerTextStyle,
  colHeaderTextStyle,
} = style;

const {
  userIdString,
  nameString,
  listHeaderString,
} = AppStrings;

function CustomerList() {

  const dispatch = useDispatch();
  const customerData = useSelector(state => state.CustomerData.data);

  const isValidCustomer = (customer) => {
    const distance = distanceFromDublinOffice(customer);
    return distance <= AppConstants.MAX_DISTANCE;
  }

  const filteredCustomerData = (customerData && customerData
                                .filter(customer => isValidCustomer(customer))
                                .sort((a,b) => a.user_id-b.user_id)) || [];

  useEffect(() => {
    let cancel = false;
    const runEffect = async () => {      
      if (cancel) {
        return;
      }
      
      dispatch(fetchCustomerData());
               
    };
    runEffect();
  
    return () => {
      cancel = true;
      
    }
  }, []);

  const renderCustomer = ({ item }) => {
    return (
      <Customer
        data={item}
      />
    );
  }

  const renderHeader = () => {
    return (
      <View>
        <Text style={headerTextStyle}>
          { listHeaderString }
        </Text>
        <View style={rowStyle}>
          <Text style={colHeaderTextStyle}>
            {userIdString}
          </Text>
          <Text style={colHeaderTextStyle}>
            { nameString }
          </Text>
        </View>
      </View>
      
    );
  }

  return (
    <FlatList
      data={filteredCustomerData}
      keyExtractor={(_, index) => index.toString()}
      renderItem={renderCustomer}
      ListHeaderComponent={renderHeader}
    />
  );
}

export default CustomerList;