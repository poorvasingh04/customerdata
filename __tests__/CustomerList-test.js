import React from 'react';
import { FlatList } from 'react-native';
import renderer from 'react-test-renderer';

import CustomerList from '../src/views/customerList/CustomerList';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => {
  const mockSelector = jest.fn();
  mockSelector.mockReturnValue(require('./data.json'));
  
  return {
    useSelector: mockSelector,
    useDispatch: () => mockDispatch
  }
});
 
describe('--Test CustomerList component--', () => {
  let testInstance;

  beforeEach(() => {
    testInstance = renderer.create(<CustomerList />).root;
  });

  it('renders correctly', () => {
    const tree = renderer.create(<CustomerList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('test filtered data values', () => {
    const data = testInstance.findByType(FlatList).props.data;
    expect(data.length).toBe(1)
  });

});