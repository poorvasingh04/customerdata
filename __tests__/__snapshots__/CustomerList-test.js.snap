// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--Test CustomerList component-- renders correctly 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "latitude": "52.986375",
        "longitude": "-6.043701",
        "name": "Christina McArdle",
        "user_id": 12,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View>
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "marginTop": 20,
              "padding": 10,
              "textAlign": "center",
              "textDecorationLine": "underline",
            }
          }
        >
          Customers within 100 Km
        </Text>
        <View
          style={
            Object {
              "borderWidth": 0.5,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "margin": 10,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "flexGrow": 1,
                "fontSize": 18,
                "fontWeight": "bold",
                "margin": 5,
                "width": 335,
              }
            }
          >
            User ID
          </Text>
          <Text
            style={
              Object {
                "flexGrow": 1,
                "fontSize": 18,
                "fontWeight": "bold",
                "margin": 5,
                "width": 335,
              }
            }
          >
            Name
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "margin": 10,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "flexGrow": 1,
              "fontSize": 16,
              "margin": 5,
              "width": 335,
            }
          }
        >
          12
        </Text>
        <Text
          style={
            Object {
              "flexGrow": 1,
              "fontSize": 16,
              "margin": 5,
              "width": 335,
            }
          }
        >
          Christina McArdle
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
